Website Search - Search UI implementation

didoo opened the pull request:
### :pushpin: Summary

This is a follow-up of the exploration done in https://github.com/hashicorp/design-system/pull/1766 where we have cherry picked the relevant commits (also cleaned/squashed them) and then built the production-ready functionality. 

### :hammer_and_wrench: Detailed description

In this PR I have:
- see [commits here](https://github.com/hashicorp/design-system/pull/1789/commits)

⚠️ Notice: to run the application you need to add some API keys (`ALGOLIA_***`) to a `.env` file inside the `website` folder. Speak with @didoo or @Dhaulagiri to get these keys).

### 👷 Testing

Use the deployed version to test the search functionality (if you want, you can test it also on mobile viewport/environments):

https://hds-website-git-website-search-search-ui-imple-e748ea-hashicorp.vercel.app/

If you don't know what to search, here are some suggestions: https://hashicorp.atlassian.net/browse/HDS-2847

### :link: External links

Jira tickets:
- https://hashicorp.atlassian.net/browse/HDS-2082 (**Epic**)

***

### 👀 Component checklist

:speech_balloon: Please consider using [conventional comments](https://conventionalcomments.org/) when reviewing this PR.

Dhaulagiri commented on website/ember-cli-build.js:
```diff
@@ -7,47 +7,54 @@
+    // override defaults to also include json files which our markdown is compiled to. without this images don't render properly.
+    replaceExtensions: ['html', 'css', 'js', 'json'],
+    // in the algolia index we want to store the fingerprinted illustrations' paths only in "production"
+    exclude: isVercelProduction ? [] : ['assets/illustrations/**'],
```
follow up question from this morning, but is there any benefit to fingerprinting these in preview apps (but not production or dev) or would it be cleaner to disable fingerprinting across the board until/unless we figure out some way of making this work in algolia (which _may_ be possible)

Dhaulagiri replied:
> may be possible

thinking out loud here that this could work if:
* Assets are fingerprinted in production
* The index is re-built based on those assets so that the fingerprinted URL in algolia matches what is in the deployment

For non-production environments we could disable fingerprinting so these assets work consistently across local dev and preview apps.

didoo commented on website/app/components/doc/page/header/index.hbs:
```diff
@@ -14,27 +14,40 @@
-    {{/if}}
-  </button>
+  <div class="doc-page-header__mobile-only-menu">
+    <button
```
This is a "fake" button that triggers the opening of the autocomplete modal (it's the only way to do it, unfortunately)

KristinLBradley replied:
Would you perhaps want to use similar markup for the mobile only nav items as for the non-mobile ones? IE, use a `nav` vs. a `div` to contain the actions and mark them up as a list. (Or perhaps move them inside the existing `nav` element and then hide and show separate the separate actions for mobile vs. non-mobile?)

alex-ju replied:
In relation to the automated a11y test failing on the `.aa-Autocomplete`. I'm wondering why don't we use this button (currently only for mobile) for the desktop version too. The markup is compliant and this way we avoid having to set the axe exception across our tests.

didoo replied:
@alex-ju in any case we need to have the `.aa-Autocomplete` injected, for the other "trigger" buttons to work. In that case, should the `.aa-Automplete` simply have `display: none` ?

didoo commented on website/app/components/doc/page/header/search/index.hbs:
```diff
@@ -0,0 +1,6 @@
+  SPDX-License-Identifier: MPL-2.0
+}}
+
+<div class="doc-algolia-search__autocomplete-container" {{did-insert this.didInsertSearchContainer}}></div>
```
This is just a container in which Algolia Autocomplete injects the "trigger" for the search modal (DOM is built by the library, and has very limited possible customization).

We add all our logic via the `didInsertSearchContainer` function invoked once the element is inserted in the DOM.

didoo commented on website/app/components/doc/page/header/search/index.js:
```diff
@@ -0,0 +1,207 @@
+
+import { getItemsFunction } from './parts/getItemsFunction';
+import { templatesHeaderFunction } from './parts/templatesHeaderFunction';
+// import { templatesFooterFunction } from './parts/templatesFooterFunction';
```
leaving this for when we will add a "show-all" link at the end of a list of results

KristinLBradley replied:
What will the "show-all" link do? Take users to a page listing the search results?

didoo replied:
I don't know yet, likely will increase the limit from 5 results to infinite for that specific type of content. 
In any case, it will be done in https://hashicorp.atlassian.net/browse/HDS-2888

didoo commented on website/app/components/doc/page/header/search/index.js:
```diff
@@ -0,0 +1,207 @@
+      // see: https://www.algolia.com/doc/ui-libraries/autocomplete/api-reference/autocomplete-js/autocomplete/#param-getsources
+      getSources({ query }) {
+        if (query.trim() !== '') {
+          return [
```
the `getSources` is used to return three different "queries" to the API (called in a single HTTP request) with different parameters so we can have different types of results (general documentation - aka "content" - and tokens and icons)

didoo commented on website/app/components/doc/page/header/search/index.js:
```diff
@@ -0,0 +1,207 @@
+            },
+          ];
+        } else {
+          return [
```
this is shown when the query is empty. we can customize what we want to show, in case

didoo commented on website/app/components/doc/page/header/search/index.js:
```diff
@@ -0,0 +1,207 @@
+
+    // add event listener to open/close the modal via `command-K` hot-key
+    document.addEventListener('keydown', (event) => {
+      if (event.metaKey && event.key === 'k') {
```
I'll finalize the command to use here later (https://hashicorp.atlassian.net/browse/HDS-2879)

didoo commented on website/ember-cli-build.js:
```diff
@@ -7,43 +7,50 @@
 
 const EmberApp = require('ember-cli/lib/broccoli/ember-app');
 
-module.exports = function (defaults) {
```
made this change to allow having conditional logic for the `fingerprint.exclude` setting.

didoo commented on website/app/styles/pages/application/header.scss:
```diff
@@ -150,15 +196,4 @@
   &:hover {
     color: var(--doc-color-white);
   }
-
```
simplified the breakpoints logic, it was used only for the homepage/header, now everything has standard media queries.

didoo commented on website/app/styles/doc-components/algolia-search/index.scss:
```diff
@@ -0,0 +1,17 @@
+// .doc-algolia-search__autocomplete-container {}
+
+// root container of the "autocomplete" injected DOM elements (only the button, in "detached" mode)
+// .aa-Autocomplete {}
```
notice: to make it easier to debug and to reference the original styles, I've kept the original naming convention: https://github.com/algolia/autocomplete/blob/next/packages/autocomplete-theme-classic/src/theme.scss

(it's impossible to completely use a custom naming convention, so better to avoid mix & match of two naming conventions)

KristinLBradley replied:
Would we perhaps want to scope the class names as a precaution still? (IE: Add a unique "hds-blah" class name to the outer container then nest other class name references inside so they don't leak.)

Just in the off chance this was ever reused elsewhere such as in an HDS component or even just to establish a convention which we stick to for this kind of thing (similarly to how I scoped the prism-dracula theme styles used by doc-code-block so they wouldn't leak and effect the HDS CodeBlock).

didoo commented on website/app/components/doc/token-preview/index.hbs:
```diff
@@ -0,0 +1,22 @@
+{{!
```
I've abstracted away the "Token Preview" component under the assumption that I could re-use it in the search results too. In reality is not possible, but since it likely will come handy in the future, I've left it here. Let me know if you prefer me to move it to a separate PR.

didoo commented on website/app/components/doc/page/header/search/parts/htmlTemplatesItemPreview.js:
```diff
@@ -0,0 +1,68 @@
+      } else {
+        return htmlPreviewIllustration({
+          html,
+          // TODO! create custom placeholder image?
```
@jorytindall let me know if you want to create a generic placeholder or just use this image/illustration

jorytindall replied:
I'll tackle this, while I think the one you've been using is probably ok, its a bit small and could be simplified a bit to "branded"

didoo commented on website/app/components/doc/page/header/search/parts/htmlTemplatesItemBody.js:
```diff
@@ -0,0 +1,91 @@
+  <div class="aa-ItemContentDescription">${description}</div>
+`;
+
+const snippetInitialText = (text) => {
```
we use this trick to return only the first 5 words of a "snippet part" (the non highlighted one) so that the highlighted content is always visible in the description field of the results in the modal panel.

KristinLBradley replied:
Doesn't it return the last five words vs. first five?

didoo replied:
the last 5 words of the _first_ snippet part (it's broken in multiple parts)

KristinLBradley commented:
I tested searching using the suggested terms [in this ticket](https://hashicorp.atlassian.net/browse/HDS-2847).

When I search for the word "color", it brings up a list of typography tokens that don't seem related to color. Is that expected?

<img width="692" alt="Screenshot 2023-11-22 at 11 30 32 AM" src="https://github.com/hashicorp/design-system/assets/108769823/f993e46b-b002-4d87-ae0c-a21ba8b96b8e">

KristinLBradley commented:
Starting to see broken images for component related links:

<img width="696" alt="image" src="https://github.com/hashicorp/design-system/assets/108769823/b48d5015-7a4f-44bf-867f-f6ff346dc38d">

KristinLBradley commented on website/app/components/doc/page/header/search/index.js:
```diff
@@ -0,0 +1,204 @@
+import { htmlPanelFooter } from './parts/htmlPanelFooter';
+import { htmlPanelNoResults } from './parts/htmlPanelNoResults';
+
+export default class DocAlgoliaSearchComponent extends Component {
```
Would we want to include the name of the tool used in the component name like this ("Algolia")?

didoo replied:
In this case yes, it's very specific and it would be a super minor thing to change it, compared with the effort to use another tool. It's to avoid confusion with a very generic term like "search" (that in some cases is used to mean "filter", hence https://github.com/hashicorp/design-system/pull/1818/files)

KristinLBradley commented on website/app/components/doc/page/header/search/index.js:
```diff
@@ -0,0 +1,204 @@
+                  attributesToSnippet: ['content:20'],
+                  distinct: true,
+                  // attributeForDistinct: 'pageBaseURL', // this doesn't work, for some reason; it needs to be set at Algolia dashboard level under "Deduplication and Grouping" configuration
+                },
```
In the documentation I found it says `attributeForDistinct` can be used in the `setSettings` method so maybe it's not supposed to work here? 

https://www.algolia.com/doc/api-reference/api-parameters/attributeForDistinct/

didoo replied:
@KristinLBradley We're using the `getAlgoliaResults` function (docs [here](https://www.algolia.com/doc/ui-libraries/autocomplete/api-reference/autocomplete-js/getAlgoliaResults/)) which points to the [Search API Parameters](https://www.algolia.com/doc/api-reference/search-api-parameters/] which exposes the `[distinct](https://www.algolia.com/doc/api-reference/api-parameters/distinct/)` parameter, but strangely enough not the `attributeForDistinct` (strangely because they both appear together in the Algolia dashboard settings)
<img width="1468" alt="screenshot_3298" src="https://github.com/hashicorp/design-system/assets/686239/73163dac-5e77-487c-812f-b378feb10532">

KristinLBradley commented on website/app/components/doc/page/header/search/parts/htmlPanelFooter.js:
```diff
@@ -0,0 +1,48 @@
+      <kbd class="esc">Esc</kbd>
+      <span>to close</span>
+    </div>
+  </div>
```
Personally, I would mark this up as a list of hints (`ul`).

KristinLBradley commented on website/app/components/doc/page/header/search/parts/htmlPanelFooter.js:
```diff
@@ -0,0 +1,48 @@
+        alt="Algolia logo"
+      />
+    </a>
+  </div>
```
And I would mark this up a a `p` vs. just a `div`.

KristinLBradley commented on website/app/components/doc/page/header/search/parts/htmlPanelFooter.js:
```diff
@@ -0,0 +1,48 @@
+      <img
+        class="aa-PanelFooter__algolia-logo"
+        src="/assets/logos/algolia.svg"
+        alt="Algolia logo"
```
The full phrase that sighted users get from this is, "Search by algolia" so the alt text should say "Algolia" vs. telling users that it's a logo which makes the phrase be "Search by Algolia logo" which doesn't make as much sense.

KristinLBradley commented on website/app/components/doc/page/header/search/parts/htmlTemplatesItemBody.js:
```diff
@@ -0,0 +1,91 @@
+
+const snippetInitialText = (text) => {
+  var splitted = text.split(' ');
+  var reduced = splitted.slice(-5);
```
Is there a reason to use `var` in here vs. `const`?

jorytindall commented:
High level note about keyboard navigation: when the modal is opened, focus is automatically set on the input, however, if I tab through the entire list of results focus proceeds to the Algolia logo, then back to the main page. This seems like a pretty significant problem, is there any control over this? Attached a screen recording.

https://github.com/hashicorp/design-system/assets/2200899/ef03d58c-6fca-4bf4-b5bf-ff6417f237fc

On an additional note about keyboard navigation, I know the up/down arrow key navigation was removed, but I can't quite remember why. I tabbed through the results which was fine, but arrow keys seem a bit more intuitive.

alex-ju reviewed (approved):
Looks good for a first iteration. I've not checked design parity (don't know where to look). Outstanding for me are the missing preview images (I know you're aware of this) and the fact that the focus is not trapped within the modal (as Jory mentioned above). However, there's a lot of code in here (1551 lines atm 🙀 ) so I'm approving it in case you want to get it in and iterate further in separate PRs

MelSumner commented:
Also, checking the page with the axe browser extension: 
![CleanShot 2023-11-27 at 10 19 19](https://github.com/hashicorp/design-system/assets/4587451/62337d10-bab2-42d5-a77a-8e43934a66ca)

didoo commented:
> Here are some issues I've found while reviewing with a screen reader:
> 
> 1. Focus is not trapped inside of the search modal.

@MelSumner this is a known issue of the library we're using https://github.com/algolia/autocomplete/issues/588. According to their comment, it's in their plan to implement it, but I don't know if/when this will really happen (and I don't think is possible to add it ourselves).

> 2. there are no instructions read to the user so I assume that I use it the way I use a normal form. I type a search term and press enter, and I expect to be taken to the list of search results. However, in pressing ENTER/RETURN, the window closes entirely and I have no idea why.

It's how it's implemented in the Algolia autocomplete library. I've pushed a hacky fix for this: https://github.com/hashicorp/design-system/pull/1789/commits/5aebac28c7a84c11889508c9876417fb9a87bd45 not super elegant and it flashes for a few milliseconds, but it's the best we can do.

> 3. Check the names/labels that are put on the items for redundancy. if the item has a role of search, then it doesn't need "search" in the accessible name.

Can you be more specific? The only element with a `role="search"` is the `<form>` element, and I don't see elements with `Search` as accessible name:

<img width="1355" alt="screenshot_3297" src="https://github.com/hashicorp/design-system/assets/686239/17652d46-6e17-41e8-8dc2-0791198b7863">

> Also, checking the page with the axe browser extension: ![CleanShot 2023-11-27 at 10 19 19](https://private-user-images.githubusercontent.com/4587451/285938499-62337d10-bab2-42d5-a77a-8e43934a66ca.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDExNjU0ODgsIm5iZiI6MTcwMTE2NTE4OCwicGF0aCI6Ii80NTg3NDUxLzI4NTkzODQ5OS02MjMzN2QxMC1iYWIyLTQyZDUtYTc3YS04ZTQzOTM0YTY2Y2EucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQUlXTkpZQVg0Q1NWRUg1M0ElMkYyMDIzMTEyOCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyMzExMjhUMDk1MzA4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NTFhNjk0YWE4MTRhYTFlNWU2YTNlMmUzNDZiOTUwZjI1M2YxODg2ZmViNDY1ZDZiY2JhZDcwZTg4ZTNlYTYxZCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.Fr-K-R74XfbNLr9f6Ecw2MD-yOaDNSSpfhAfjEpbRKM)

Yes, that's the reason why I opened https://github.com/hashicorp/design-system/pull/1823. Unfortunately there is no way to fix it, the library doesn't provide overrides or hooks that we can use. In theory we could try to open an upstream PR, but I am not sure it would be easy (it's a very complex library, with a lot of other libraries depending on it).

KristinLBradley commented on website/app/styles/doc-components/algolia-search/modal.scss:
```diff
@@ -0,0 +1,250 @@
+
+  @include breakpoint.with-fixed-sidebar () {
+    position: absolute;
+    top: 90px;
```
If this `top: 90px;` value is based partly on the height of the doc-page-header would it make sense to use a variable or else add a comment explaining?

didoo replied:
It's kinda base, but not exactly. It's a number that "looks right" (visually, and numerically).

KristinLBradley commented on website/app/styles/doc-components/algolia-search/modal.scss:
```diff
@@ -0,0 +1,250 @@
+
+  &::after {
+    position: absolute;
+    top: -16px;
```
I prefer to use `bottom: 100%;` for positioning items like this. That way if the height were ever adjusted for some reason, you don't have to remember to change a value in 2 places.

didoo replied:
You're absolutely right. Fixed! 🙏

KristinLBradley commented on website/app/styles/doc-components/algolia-search/modal.scss:
```diff
@@ -0,0 +1,250 @@
+    color: var(--doc-color-gray-200);
+    font-weight: 500;
+    font-size: 13px;
+    line-height: 14px;
```
If you change this to `line-height: 1;` and add: `display: flex; align-items: center;` then you don't need a fixed value like this for the line-height. (Also if the height of the button and/or text size is ever adjusted then it will still be vertically centered.)

KristinLBradley commented on website/app/styles/doc-components/algolia-search/modal.scss:
```diff
@@ -0,0 +1,250 @@
+    &.tab,
+    &.esc {
+      font-size: 11px;
+      letter-spacing: -0.6px;
```
It you use ems for letter-spacing values then it adjusts if the font-size is changed for any reason since it's relative to the font size instead of being a fixed value. (Try changing to: `letter-spacing: -.05em;`)

KristinLBradley commented on website/app/styles/doc-components/algolia-search/trigger.scss:
```diff
@@ -0,0 +1,60 @@
+  margin: 0;
+  padding: 4px 12px;
+  color: var(--doc-color-gray-400);
+  line-height: 16px; // override
```
I didn't see any effect when I commented out this line-height value in the browser

didoo replied:
Removed.

KristinLBradley reviewed (commented):
Looking great overall! Went through and tested functionality & looked through code in detail. Left some comments & questions.

didoo commented:
> When I search for the word "color", it brings up a list of typography tokens that don't seem related to color. Is that expected?

@KristinLBradley It's because the tokens for the font families contains the text `Apple *Color* Emoji` in their values. Something I need to discuss with @jorytindall 

> Would it be possible to display a message saying "no exact matches for 'blah' " or similar followed by a list of similar word matches found?

I think it may be possible, but not out of the box (it would require some extra custom logic).

didoo commented:
> High level note about keyboard navigation: when the modal is opened, focus is automatically set on the input, however, if I tab through the entire list of results focus proceeds to the Algolia logo, then back to the main page. This seems like a pretty significant problem, is there any control over this? Attached a screen recording.

@jorytindall Unfortunately no. There is an open issue here https://github.com/algolia/autocomplete/issues/588 but is not being picked up apparently.

> On an additional note about keyboard navigation, I know the up/down arrow key navigation was removed, but I can't quite remember why. I tabbed through the results which was fine, but arrow keys seem a bit more intuitive.

It's not been removed. It works only when the focus is on the `input` element, but not when an element is focused. Same as above, I have no control on this (that's why I decided to hide the tip in the modal's footer).

didoo commented:
@alex-ju @jorytindall @KristinLBradley @MelSumner I've addressed most of the comments/suggestions and for the ones that I couldn't address I've explained why (see comments above).

At this point I think the question to answer is: is it good enough to go live in production as "beta"? if the answer is "yes" then probably is ok to approve it. In any case we have a [list of follow-up tasks](https://hashicorp.atlassian.net/browse/HDS-2862) that will be worked on later anyway.

You can test the search functionality here: https://hds-website-git-website-search-search-ui-imple-e748ea-hashicorp.vercel.app/

/cc @Dhaulagiri

didoo commented on website/app/components/doc/page/header/index.hbs:
```diff
@@ -14,27 +14,52 @@
       <Doc::Page::Header::NavItem @label="Components" @route="components" @currentTopRoute={{@currentTopRoute}} />
       <Doc::Page::Header::NavItem @label="Patterns" @route="patterns" @currentTopRoute={{@currentTopRoute}} />
+      <li class="doc-page-header__nav-item-generic doc-page-header__nav-item--split">
+        <div class="sr-only" id="search-button">This button opens a dialog containing an input field. Entering text in the input will perform -automatically as you type- a search within the website.</div>
```
@alex-ju good point. done in https://github.com/hashicorp/design-system/pull/1848

MelSumner commented on website/app/components/doc/page/header/index.hbs:
```diff
@@ -14,27 +14,52 @@
       <Doc::Page::Header::NavItem @label="Components" @route="components" @currentTopRoute={{@currentTopRoute}} />
       <Doc::Page::Header::NavItem @label="Patterns" @route="patterns" @currentTopRoute={{@currentTopRoute}} />
+      <li class="doc-page-header__nav-item-generic doc-page-header__nav-item--split">
+        <div class="sr-only" id="search-button">This button opens a dialog containing an input field. Entering text in the input will perform -automatically as you type- a search within the website.</div>
```
Suggestion: 
```suggestion
        <div class="sr-only" id="search-button">This button opens a dialog containing an input field and some additional information that you may wish to explore. An automatic search will be performed as you type text into the search field, but the results may not be announced. Exploring the additional items in the modal will help you discover the search results.</div>
```

MelSumner commented on website/app/components/doc/page/header/index.hbs:
```diff
@@ -14,27 +14,52 @@
+          type="button"
+          class="doc-page-header__search-button"
+          data-doc-algolia-search-autocomplete-secondary-trigger
+          aria-describedby="search-button"
```
👍

didoo commented:
> Not referencing the additions to the UI we discussed today

@jorytindall I've pushed the change we discussed (the "See all..." link at the bottom of the results lists for icons and tokens). Let me know what you think and if you would do some changes (font size or weight? color? add an icon? other?)

MelSumner reviewed (approved):
Let's ship it.
